### PR TITLE
feat(mr): create namespace for Model Registry

### PR DESCRIPTION
**Pull request by zdtsw**
Create dedicated namespace for Model registry to use.
Current this namespace will be used for ODH and downstream but we will have a followup PR to set the correct namespace and logic only for downstream

ref https://issues.redhat.com/browse/RHOAIENG-5027

## Description

- create a NS for MR once the component is enabled
- this NS wont get deleted even MR component is set to Removed/Unmanagaedor ODH operator is being uninstalled

## How Has This Been Tested?

result to a namespace with similar labels:
```
labels:
    kubernetes.io/metadata.name: odh-model-registries
    ...
```

## Merge criteria:

- [ ] The commits are squashed in a cohesive manner and have meaningful messages.
- [ ] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [ ] The developer has manually tested the changes and verified that the changes work

### CODE REVIEW

**Comment by zdtsw** [comment]
@dhirajsb wanna check if "odh-model-registries" is the right namespace you want to use (from our discussion yesterday my understanding is we will not have a configable namespace for now, which should have a static name)

**Comment by dhirajsb** [comment]
@bartoszmajsak can we add a label/annotation here that will automatically create a servicemeshmember to add this namespace in the odh control plane as well?

**Comment by zdtsw** [comment]
> @bartoszmajsak can we add a label/annotation here that will automatically create a servicemeshmember to add this namespace in the odh control plane as well?

another way to create this namespace is going thro featuretracker.
we can have a precondistion of creating such resource, then the featuretracker itself will be used to create smm

**Comment by bartoszmajsak** [comment]
> @bartoszmajsak can we add a label/annotation here that will automatically create a servicemeshmember to add this namespace in the odh control plane as well?

We actually had it done this way in the first version with `KfDef` plugin and our own controller. I will get back to you this week about the approach. /cc @aslakknutsen

**Review by dhirajsb — APPROVED** [review]
/lgtm
@zdtsw thanks for the PR. We should go ahead and add the `odh-model-registries` namespace to incubation for the coming odh 2.10 release. 
We can create another PR for rhoai namespace change to `redhat-model-registries`.

**Comment by dhirajsb** [comment]
@zdtsw @bartoszmajsak let's work on adding the `smm` to this namespace in another PR. 
I created jira issues https://issues.redhat.com/browse/RHODS-13699 for namespace creation, and https://issues.redhat.com/browse/RHODS-13700 for adding smm to the namespace.

**Comment by zdtsw** [comment]
> /lgtm @zdtsw thanks for the PR. We should go ahead and add the `odh-model-registries` namespace to incubation for the coming odh 2.10 release. We can create another PR for rhoai namespace change to `redhat-model-registries`.

@VaishnaviHire if you are fine with this PR we can get it for today's code freeze.
as for the namespace for RHOAI, we can use another PR to handle it later

**Comment by bartoszmajsak** [comment]
> @zdtsw @bartoszmajsak let's work on adding the `smm` to this namespace in another PR. I created jira issues [issues.redhat.com/browse/RHODS-13699](https://issues.redhat.com/browse/RHODS-13699) for namespace creation, and [issues.redhat.com/browse/RHODS-13700](https://issues.redhat.com/browse/RHODS-13700) for adding smm to the namespace.

@zdtsw @dhirajsb I'm back in action on Tue, so let's touch on it then. 

On semi related note - are we still supposed to use RHODS JIRA? I though RHOAIENG is now the place to capture tasks :)

**Comment by dhirajsb** [comment]
I must have been looking at some old RHODS issue and hit the create button without noticing :laughing: 
I moved the issues to the Epics https://issues.redhat.com/browse/RHOAIENG-5027 and https://issues.redhat.com/browse/RHOAIENG-5028 respectively. 
I've set the release version as rhoai 2.10, but an odh 2.10.x release will do.

**Comment by dhirajsb** [comment]
@zdtsw shouldn't this fix also check whether the namespace already exists? The user could enable, remove, then enable model registry again.

File: components/modelregistry/modelregistry.go [hunk 1]
```diff
@@ -79,6 +81,12 @@ func (m *ModelRegistry) ReconcileComponent(_ context.Context, cli client.Client,
 				return err
 			}
 		}
+
+		// Create odh-model-registries namespace
+		_, err := cluster.CreateNamespace(cli, "odh-model-registries", cluster.WithLabels(labels.ODH.OwnedNamespace, "true"))
+		if err != nil {
+			return err
+		}
```
Comment (bartoszmajsak) [inline components/modelregistry/modelregistry.go:89]: We don't have [unified labels in place yet](https://github.com/bartoszmajsak/opendatahub-architecture-decision-records/blob/5df775f582393dc640cd805080184a038d2b56f0/platform/ODH-ADR-0001-k8s-labels-standard.md#L84),

but I was thinking we can establish a bit more meaningful relationship here:

```suggestion
	modelRegistryLabels := cluster.WithLabels(
		labels.ODH.OwnedNamespace, "true",
		labels.K8SCommon.PartOf, ComponentName,
		labels.ODH.Component(ComponentName), "true",
	)
	_, err := cluster.CreateNamespace(cli, "odh-model-registries", modelRegistryLabels)
	if err != nil {
		return err
	}
```

`modelRegistryLabels` can also be a top-level var, as it's static. WDYT @zdtsw @dhirajsb?

**Review by bartoszmajsak — COMMENTED** [review]
I have one minor suggestion about the labels.

File: components/modelregistry/modelregistry.go (hunk 1 above)
Comment (bartoszmajsak) [reply components/modelregistry/modelregistry.go:89]: I wonder if we shouldn't add short comment on "why" we do not set ownership for this namespace, or more generally speaking, why the operator shouldn't remove it.

So explaining this decision from the PR description, but on the code level:

> this NS wont get deleted even MR component is set to Removed/Unmanagaed or ODH operator is being uninstalled.

Comment (zdtsw) [reply components/modelregistry/modelregistry.go:89]: i am good to make modelRegistryLabels with labels, but want to discuss more regarding 2 extra/new labels

Comment (zdtsw) [reply components/modelregistry/modelregistry.go]: did update with
- use variable to catch label
- add inline comments to keep namespace

wait for some offline discussion if should add 2 new labels or not

**Comment by zdtsw** [comment]
I must correct myself:   @dhirajsb @bartoszmajsak 
by having label "opendatahub.io/generated-namespace=true" on namespace, this does get namespace deleted during uninstallation of operator.

**Comment by dhirajsb** [comment]
@zdtsw we don't want the namespace to be deleted on model-registry uninstall. Sorry, I wasn't aware that's what that label did.

**Comment by zdtsw** [comment]
> @zdtsw we don't want the namespace to be deleted on model-registry uninstall. Sorry, I wasn't aware that's what that label did.


to be more clear, here "uninstallation" from my previous comments means  ODH operator uninstallation, not to set component Model Registry to Removed.

**Review by dhirajsb — APPROVED** [review]
/lgtm

**Comment by zdtsw** [comment]
e2e failed due to CFO bug

**Review by bartoszmajsak — APPROVED** [review]

**Pull request merged**